File: dev/test_data/golang/test_main/aas_core_meta.v3/input/snippets/Verification/submodel_elements_have_identical_semantic_IDs.go
// Check that all `elements` have the identical [aastypes.IHasSemantics.SemanticID].
func SubmodelElementsHaveIdenticalSemanticIDs[S aastypes.ISubmodelElement](
	elements []S) bool {
	var thatSemanticID aastypes.IReference

	for _, element := range elements {
		thisSemanticID := element.SemanticID()

		if thisSemanticID == nil {
			continue
		}

		if thatSemanticID == nil {
			thatSemanticID = thisSemanticID
			continue
		}

		thisKeys := thisSemanticID.Keys()
		thatKeys := thatSemanticID.Keys()

		if len(thisKeys) != len(thatKeys) {
			return false
		}

		for i, thisKey := range thisKeys {
			thatKey := thatKeys[i]

			if thisKey.Value() != thatKey.Value() {
				return false
			}
		}
	}
	return true
}

File: dev/test_data/typescript/test_main/aas_core_meta.v3/input/snippets/Verification/submodel_elements_have_identical_semantic_IDs.ts
/**
 * Check that all `elements` have the identical
 * {@link types.ISubmodelElement.semanticId}.
 *
 * @param elements - to be checked
 * @returns `true` if all the semantic IDs are identical
 */
export function submodelElementsHaveIdenticalSemanticIds(
  elements: Iterable<AasTypes.ISubmodelElement>
): boolean {
  let thatSemanticId: AasTypes.Reference | null = null;
  for (const element of elements) {
    if (element.semanticId === null) {
      continue;
    }

    if (thatSemanticId === null) {
      thatSemanticId = element.semanticId;
      continue;
    }

    const thisSemanticId = element.semanticId;

    if (thisSemanticId.keys.length != thatSemanticId.keys.length) {
      return false;
    }

    for (let i = 0; i < thatSemanticId.keys.length; i++) {
      if (thisSemanticId.keys[i].value !== thatSemanticId.keys[i].value) {
        return false;
      }
    }
  }

  return true;
}
